cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClientService.java

```diff
@@ -443,4 +500,147 @@ private long findProgramFinishTime(List<Message> messages) {
       return -1L;
     }
   }
+
+  /**
+   * An aggregator for messages.
+   */
+  private interface MessageAggregator {
+    Iterator<Message> aggregate(Iterator<Message> input);
+  }
+
+  /**
+   * An aggregator for metrics that combines metrics with the same tags and metric name.
+   */
+  private static class MetricsMessageAggregator implements MessageAggregator {
+    private final HashMap<String, String> cachemap;
+    private final MessagingMetricsProcessorService.PayloadInputStream decodingOutputStream;
+    private final StringCachingDecoder decoder;
+    private final DatumReader<MetricValues> reader;
+    private final Schema schema;
+    private final ByteArrayOutputStream encodeOutputStream;
+    private final BinaryEncoder encoder;
+    private final DatumWriter<MetricValues> recordWriter;
+
+    MetricsMessageAggregator(DatumReader<MetricValues> reader, Schema schema, DatumWriter<MetricValues> recordWriter) {
+      // One time set up for decoding.
+      this.cachemap = new HashMap<>();
+      decodingOutputStream = new MessagingMetricsProcessorService.PayloadInputStream();
+      decoder = new StringCachingDecoder(new BinaryDecoder(decodingOutputStream), cachemap);
+      this.encodeOutputStream = new ByteArrayOutputStream(1024);
+      encoder = new BinaryEncoder(encodeOutputStream);
+      this.reader = reader;
+      this.schema = schema;
+      this.recordWriter = recordWriter;
+    }
+
+    @Override
+    public Iterator<Message> aggregate(Iterator<Message> input) {
+      if (!input.hasNext()) {
+        return input;
+      }
+      ArrayList<Message> messages = new ArrayList<>();
+      while (input.hasNext()) {
+        Message message = input.next();
+        messages.add(message);
+      }
+
+      ArrayList<MetricValues> decoded = new ArrayList<>();
+
+      // Decode the messages into MetricValues.
+      String lastMessageId = null;
+      for (Message message : messages) {
+        lastMessageId = message.getId();
+        decodingOutputStream.reset(message.getPayload());
+        try {
+          decoded.add(reader.read(decoder, schema));
```

Comment:
Again, why need to decode everything before aggregate? The decode + aggregation should be able to get done in a streaming fashion.

Reply:
I've made changes to do decoding and aggregation while iterating now.